src/crystal/system/unix/io_uring.cr

```diff
@@ -0,0 +1,760 @@
+  # more. This check is only needed before the NODROP feature was implemented (Linux 5.5+)
+  private def get_free_index : UInt32
+    # If there are too many in flight events, wait for some completions.
+    until @params.features.nodrop? || @inflight < @params.cq_entries
+      process_completion_events(blocking: true)
+    end
+
+    loop do
+      index = @submission_queue.consume_free_index
+
+      # If the submission queue is full, submit some events.
+      if index == UInt32::MAX
+        process_completion_events(blocking: false)
+        next
+      end
+
+      @inflight += 1
+      return index
+    end
+  end
+
+  # Obtains one submission entry, populates, and submits it. When the completion
+  # event arrives the current Fiber will be enqueued for execution.
+  # The caller of this method MUST either reschedule the current Fiber or
+  # resume into another Fiber.
+  private def submit(opcode : Syscall::IoUringOp, user_data : UInt64, *, index : UInt32 = get_free_index, timeout : ::Time::Span? = nil)
+    # The submission queue just stores indices into the `submission_entries` array.
+    sqe = @submission_entries + index
+
+    # This submission queue entry is reused! Clean it up.
+    Intrinsics.memset(sqe, 0_u8, sizeof(Syscall::IoUringSqe), false)
+
+    # Populate fields and submit the index into the submission queue.
+    yield sqe
+    sqe.value.user_data = user_data
+    sqe.value.opcode = opcode
+
+    # If there is a timeout, we submit a LINK_TIMEOUT as well. It will make the
+    # original event fail with ECANCELED if it doesn't complete in time.
+    if timeout
+      index_timeout = get_free_index
+
+      @timeval_entries[index_timeout] = make_timeval(timeout)
+
+      sqe_timeout = @submission_entries + index_timeout
+      Intrinsics.memset(sqe_timeout, 0_u8, sizeof(Syscall::IoUringSqe), false)
+      sqe_timeout.value.addr = (@timeval_entries.to_unsafe + index_timeout).address
+      sqe_timeout.value.len = 1u32
+
+      if IoUring.probe.operations.link_timeout?
+        sqe_timeout.value.opcode = Syscall::IoUringOp::LINK_TIMEOUT
+        sqe.value.flags = sqe.value.flags | Syscall::IoUringSqeFlags::IO_LINK
+      else
+        id = @next_cancelable_poll_info_id
+        @next_cancelable_poll_info_id += 1
+        @cancelable_poll_info[id] = CancelablePollInfo.new(user_data, CancelablePollStatus::Running)
+
+        sqe.value.user_data = (id << 2) + 2
+        sqe_timeout.value.opcode = Syscall::IoUringOp::TIMEOUT
+        sqe_timeout.value.user_data = (id << 2) + 3
+      end
+
+      @submission_queue.push(index)
+      @submission_queue.push(index_timeout)
+    else
+      @submission_queue.push(index)
+    end
+  end
+
+  # This will block the current fiber until the request completes and return the completion result.
+  private def submit_and_wait(opcode : Syscall::IoUringOp, *, index : UInt32 = get_free_index, timeout : ::Time::Span? = nil)
+    user_data = ::Fiber.current.as(Void*).address
```

Comment:
`::Fiber.current.object_id` might suffice.

Reply:
Is the semantic of `object_id` guaranteed to be a pointer to the object? I assume it just means "a unique number for this object" but it could be `pointer+1` without breaking semantics. As I'm casting it back from address to Fiber, I think it is more legible to make an explicit cast through pointers. Makes sense?

Reply:
It is documented to be the address, and there are other places that also depend on it. `Crystal::FiberChannel`, `Object#clone, GC.add_root`,  etc.